File: dawn_cpp/lsh.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <string>
#include <algorithm>

class LSHHash {
public:
    static int hash(const std::string& str) {
        int sum = 0;
        for (char c : str) {
            sum += static_cast<int>(c);
        }
        return sum;
    }

    static int hammingDistance(const std::string& str1, const std::string& str2) {
        int distance = 0;
        int len = std::min(str1.length(), str2.length());
        for (int i = 0; i < len; i++) {
            if (str1[i] != str2[i]) {
                distance++;
            }
        }
        return distance;
    }

    static std::vector<int> lshFunc(int k, int l, const std::string& str) {
        std::vector<int> hashes(k);
        int len = str.length();
        for (int i = 0; i < k; i++) {
            int start = i * len / k;
            int end = start + l * len / k;
            hashes[i] = hash(str.substr(start, end - start));
        }
        return hashes;
    }

    static std::unordered_map<int, std::vector<std::string>> buildLSHTable(int k, int l, const std::vector<std::string>& strings) {
        std::unordered_map<int, std::vector<std::string>> table;
        for (const auto& str : strings) {
            std::vector<int> hashes = lshFunc(k, l, str);
            for (int hashVal : hashes) {
                table[hashVal].push_back(str);
            }
        }
        return table;
    }

    static std::vector<std::string> queryLSHTable(int k, int l, const std::unordered_map<int, std::vector<std::string>>& table, const std::string& queryStr) {
        std::vector<std::string> candidates;
        std::vector<int> hashes = lshFunc(k, l, queryStr);
        for (int hashVal : hashes) {
            auto it = table.find(hashVal);
            if (it != table.end()) {
                for (const auto& str : it->second) {
                    if (hammingDistance(str, queryStr) <= l) {
                        candidates.push_back(str);
                    }
                }
            }
        }
        return candidates;
    }
};

File: dawn_rust/src/lsh.rs
use std::collections::HashMap;

pub fn hash(str: &str) -> u32 {
    str.chars().map(|c| c as u32).sum()
}

pub fn hamming_distance(str1: &str, str2: &str) -> u32 {
    str1.chars()
        .zip(str2.chars())
        .filter(|&(c1, c2)| c1 != c2)
        .count() as u32
}

pub fn lsh_func(k: usize, l: usize, str: &str) -> Vec<u32> {
    let len = str.len();
    let mut hashes = Vec::with_capacity(k);
    for i in 0..k {
        let start = i * len / k;
        let end = start + l * len / k;
        let hash_val = hash(&str[start..end]);
        hashes.push(hash_val);
    }
    hashes
}

pub fn build_lsh_table(k: usize, l: usize, strings: &[&str]) -> HashMap<u32, Vec<&str>> {
    let mut table = HashMap::new();
    for &str in strings {
        let hashes = lsh_func(k, l, str);
        for hash_val in hashes {
            table.entry(hash_val).or_insert_with(Vec::new).push(str);
        }
    }
    table
}

pub fn query_lsh_table(k: usize, l: usize, table: &HashMap<u32, Vec<&str>>, query_str: &str) -> Vec<&str> {
    let hashes = lsh_func(k, l, query_str);
    let mut candidates = Vec::new();
    for hash_val in hashes {
        if let Some(bucket) = table.get(&hash_val) {
            for &str in bucket {
                if hamming_distance(str, query_str) <= l as u32 {
                    candidates.push(str);
                }
            }
        }
    }
    candidates
}